File: apps/desktop/src/api/tauri/tasks.ts
// @author kongweiguang
// SPDX-License-Identifier: GPL-3.0-or-later

import { z } from "zod";
import {
  parseMethodParams,
  parseMethodResult,
  type MethodParams,
  type MethodResult,
} from "../protocol/methods";
import {
  defaultNativeBridge,
  normalizeRuntimeError,
  RuntimeHostError,
  type RuntimeNativeBridge,
} from "./runtime";

export const JA_TASK_COMMANDS = {
  create: "ja_runtime_task_create",
  list: "ja_runtime_task_list",
  read: "ja_runtime_task_read",
  observe: "ja_runtime_task_observe",
  unobserve: "ja_runtime_task_unobserve",
  seen: "ja_runtime_task_seen",
  messageSend: "ja_runtime_task_message_send",
  followup: "ja_runtime_task_followup",
  cancel: "ja_runtime_task_cancel",
  treeDelete: "ja_runtime_task_tree_delete",
} as const;

export type TaskCreateInput = MethodParams<"task/create">;
export type TaskCreateResult = MethodResult<"task/create">;
export type TaskListInput = MethodParams<"task/list">;
export type TaskListResult = MethodResult<"task/list">;
export type TaskReadInput = MethodParams<"task/read">;
export type TaskReadResult = MethodResult<"task/read">;
export type TaskObserveInput = MethodParams<"task/observe">;
export type TaskObserveResult = MethodResult<"task/observe">;
export type TaskUnobserveInput = MethodParams<"task/unobserve">;
export type TaskSeenInput = MethodParams<"task/seen">;
export type TaskMessageInput = MethodParams<"task/message/send">;
export type TaskMessageResult = MethodResult<"task/message/send">;
export type TaskFollowupInput = MethodParams<"task/followup">;
export type TaskFollowupResult = MethodResult<"task/followup">;
export type TaskMutationInput = MethodParams<"task/cancel">;
export type TaskMutationResult = MethodResult<"task/cancel">;
export type TaskTreeDeleteInput = MethodParams<"task/tree/delete">;
export type TaskTreeDeleteResult = MethodResult<"task/tree/delete">;

type TaskMethod =
  | "task/create"
  | "task/list"
  | "task/read"
  | "task/observe"
  | "task/unobserve"
  | "task/seen"
  | "task/message/send"
  | "task/followup"
  | "task/cancel"
  | "task/tree/delete";

export interface TaskAdapter {
  create(input: TaskCreateInput): Promise<TaskCreateResult>;
  list(input: TaskListInput): Promise<TaskListResult>;
  read(input: TaskReadInput): Promise<TaskReadResult>;
  observe(input: TaskObserveInput): Promise<TaskObserveResult>;
  unobserve(input: TaskUnobserveInput): Promise<void>;
  seen(input: TaskSeenInput): Promise<TaskMutationResult>;
  messageSend(input: TaskMessageInput): Promise<TaskMessageResult>;
  followup(input: TaskFollowupInput): Promise<TaskFollowupResult>;
  cancel(input: TaskMutationInput): Promise<TaskMutationResult>;
  treeDelete(input: TaskTreeDeleteInput): Promise<TaskTreeDeleteResult>;
}

/**
 * 每个产品动作绑定一个专用 Tauri command；method 只在 adapter 内部选择，Renderer 调用方
 * 无法把这个边界降级为通用 JA-RPC tunnel。
 */
async function invokeTask<M extends TaskMethod>(
  bridge: RuntimeNativeBridge,
  command: (typeof JA_TASK_COMMANDS)[keyof typeof JA_TASK_COMMANDS],
  method: M,
  input: MethodParams<M>,
): Promise<MethodResult<M>> {
  let parsed: MethodParams<M>;
  try {
    parsed = parseMethodParams(method, input);
  } catch {
    throw new RuntimeHostError("INVALID_INPUT", "请求参数无效", false);
  }
  try {
    const result = await bridge.invoke<unknown>(command, { input: parsed });
    return parseMethodResult(method, result);
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new RuntimeHostError("RUNTIME_UNAVAILABLE", "运行时暂不可用", true);
    }
    throw normalizeRuntimeError(error);
  }
}

/** 专用 Task adapter 保持 observe/unobserve 对称，并把所有结果留在 canonical Schema 门内。 */
export class TauriTaskAdapter implements TaskAdapter {
  /** bridge 仅供合同测试注入；生产始终使用 Runtime 的单一原生边界。 */
  constructor(private readonly bridge: RuntimeNativeBridge = defaultNativeBridge) {}

  /** 侧边任务只通过用户专用 create command 创建，不能从 UI 冒充 Subagent。 */
  create(input: TaskCreateInput): Promise<TaskCreateResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.create, "task/create", input);
  }

  /** 总览只返回根任务的有界摘要，不物化任何 Child Transcript。 */
  list(input: TaskListInput): Promise<TaskListResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.list, "task/list", input);
  }

  /** 详情元数据与 Mailbox 使用有界分页读取，正文继续由 thread/read 独立负责。 */
  read(input: TaskReadInput): Promise<TaskReadResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.read, "task/read", input);
  }

  /** 高频观察必须携带 revision CAS，防止详情切换后订阅陈旧实例。 */
  observe(input: TaskObserveInput): Promise<TaskObserveResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.observe, "task/observe", input);
  }

  /** 关闭详情只释放观察句柄，不取消正在执行的任务。 */
  async unobserve(input: TaskUnobserveInput): Promise<void> {
    await invokeTask(this.bridge, JA_TASK_COMMANDS.unobserve, "task/unobserve", input);
  }

  /** 已读边界由服务端 sequence 与 revision CAS 决定，UI 不本地猜测消费进度。 */
  seen(input: TaskSeenInput): Promise<TaskMutationResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.seen, "task/seen", input);
  }

  /** QueueOnly 消息不会在目标空闲时隐式启动新 Turn。 */
  messageSend(input: TaskMessageInput): Promise<TaskMessageResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.messageSend, "task/message/send", input);
  }

  /** follow-up 是唯一允许从详情显式启动或排队 Child Turn 的交互。 */
  followup(input: TaskFollowupInput): Promise<TaskFollowupResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.followup, "task/followup", input);
  }

  /** 取消提交传播意图；最终状态仍以持久事件和重读投影为准。 */
  cancel(input: TaskMutationInput): Promise<TaskMutationResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.cancel, "task/cancel", input);
  }

  /** 整树删除要求双重 identity，避免普通关闭 Tab 或删除 Thread 产生数据破坏。 */
  treeDelete(input: TaskTreeDeleteInput): Promise<TaskTreeDeleteResult> {
    return invokeTask(this.bridge, JA_TASK_COMMANDS.treeDelete, "task/tree/delete", input);
  }
}

/** 生产工厂只返回窄 Task surface，调用方接触不到 bridge 或 command 名。 */
export function createTaskAdapter(): TaskAdapter {
  return new TauriTaskAdapter();
}

File: src-tauri/src/app_runtime/domain/tasks.rs
// SPDX-License-Identifier: GPL-3.0-or-later
// @author kongweiguang

//! Child Thread 任务协议的纯领域输入与权威投影。
//!
//! Rust 只验证 WebView 可提交的闭集并承载 Java 返回的安全投影，不保存任务树、
//! Mailbox 或 observation 的第二份权威状态。

use super::commands::{DomainValidationError, TurnContentPart, validate_turn_content};
use super::{valid_frozen_turn_id, valid_protocol_id};

const MAX_SAFE_JSON_INTEGER: u64 = 9_007_199_254_740_991;

#[derive(Debug, Clone)]
pub struct TaskCreateInput {
    pub parent_thread_id: String,
    pub parent_turn_id: Option<String>,
    pub expected_parent_revision: u64,
    pub task_name: String,
    pub content: Vec<TurnContentPart>,
}

impl TaskCreateInput {
    /// 在进入 sidecar actor 前关闭 task/create 的字段与大小边界；父子关系和 revision
    /// 仍由 Java 在同一事务裁决，Rust 不建立本地任务树。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        if !valid_protocol_id(&self.parent_thread_id, "thr_", 100)
            || self
                .parent_turn_id
                .as_deref()
                .is_some_and(|turn_id| !valid_frozen_turn_id(turn_id))
            || self.expected_parent_revision > MAX_SAFE_JSON_INTEGER
            || self.task_name.trim() != self.task_name
            || self.task_name.is_empty()
            || self.task_name.chars().count() > 96
            || self.task_name.chars().any(char::is_control)
        {
            return Err(DomainValidationError);
        }
        validate_turn_content(&self.content, 4_000_000)
    }
}

#[derive(Debug, Clone)]
pub struct TaskListInput {
    pub root_thread_id: String,
}

impl TaskListInput {
    /// task/list 只接受树根 identity，避免 renderer 请求全局任务扫描或 Child transcript。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        valid_protocol_id(&self.root_thread_id, "thr_", 100)
            .then_some(())
            .ok_or(DomainValidationError)
    }
}

#[derive(Debug, Clone)]
pub struct TaskReadInput {
    pub task_thread_id: String,
    pub cursor: Option<String>,
    pub limit: Option<u16>,
}

impl TaskReadInput {
    /// 读取任务详情只接受活动/Mailbox 双序号 cursor，防止畸形 opaque 值绕过服务端分页边界。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        if !valid_protocol_id(&self.task_thread_id, "thr_", 100)
            || self.limit.is_some_and(|limit| !(1..=200).contains(&limit))
            || self
                .cursor
                .as_deref()
                .is_some_and(|cursor| !valid_task_cursor(cursor))
        {
            return Err(DomainValidationError);
        }
        Ok(())
    }
}

/// Task cursor 是 `task:<activitySequence>:<mailboxSequence>`；两段十进制允许 0 表示第一页边界。
pub(crate) fn valid_task_cursor(value: &str) -> bool {
    if value.len() > 256 {
        return false;
    }
    let mut parts = value.split(':');
    matches!(parts.next(), Some("task"))
        && parts.next().is_some_and(valid_cursor_sequence)
        && parts.next().is_some_and(valid_cursor_sequence)
        && parts.next().is_none()
}

/// Cursor sequence 禁止空串、符号和超过 JavaScript safe integer 的值。
fn valid_cursor_sequence(value: &str) -> bool {
    !value.is_empty()
        && value.bytes().all(|byte| byte.is_ascii_digit())
        && value
            .parse::<u64>()
            .is_ok_and(|number| number <= MAX_SAFE_JSON_INTEGER)
}

#[derive(Debug, Clone)]
pub struct TaskObserveInput {
    pub task_thread_id: String,
    pub expected_task_revision: u64,
}

impl TaskObserveInput {
    /// observation 绑定 Java projection revision；Rust 不从已见事件推断或修正该 CAS。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        validate_task_revision(&self.task_thread_id, self.expected_task_revision)
    }
}

#[derive(Debug, Clone)]
pub struct TaskUnobserveInput {
    pub observation_id: String,
}

impl TaskUnobserveInput {
    /// 释放只接受 App Server 签发的 connection-scoped handle，禁止把 Thread ID 当 handle。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        valid_protocol_id(&self.observation_id, "observe_", 103)
            .then_some(())
            .ok_or(DomainValidationError)
    }
}

#[derive(Debug, Clone)]
pub struct TaskSeenInput {
    pub task_thread_id: String,
    pub expected_task_revision: u64,
    pub through_activity_sequence: u64,
}

impl TaskSeenInput {
    /// 已读边界必须和 projection revision 一起提交，避免迟到详情覆盖更新后的未读事实。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        validate_task_revision(&self.task_thread_id, self.expected_task_revision)?;
        if !(1..=MAX_SAFE_JSON_INTEGER).contains(&self.through_activity_sequence) {
            return Err(DomainValidationError);
        }
        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct TaskMessageInput {
    pub sender_thread_id: String,
    pub target_thread_id: String,
    pub content: Vec<TurnContentPart>,
    pub idempotency_key: String,
}

impl TaskMessageInput {
    /// Mailbox 写入只接收结构化内容与有界幂等键；关系授权和 exactly-once 由 Java 事务负责。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        if !valid_protocol_id(&self.sender_thread_id, "thr_", 100)
            || !valid_protocol_id(&self.target_thread_id, "thr_", 100)
            || self.idempotency_key.is_empty()
            || self.idempotency_key.len() > 128
            || self.idempotency_key.chars().any(char::is_control)
        {
            return Err(DomainValidationError);
        }
        validate_turn_content(&self.content, 524_288)
    }
}

#[derive(Debug, Clone)]
pub struct TaskFollowupInput {
    pub message: TaskMessageInput,
    pub expected_task_revision: u64,
}

impl TaskFollowupInput {
    /// followup 与 QueueOnly message 保持同一内容边界，但额外要求目标 revision 以启动 Child Turn。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        self.message.validate()?;
        if self.expected_task_revision > MAX_SAFE_JSON_INTEGER {
            return Err(DomainValidationError);
        }
        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct TaskMutationInput {
    pub task_thread_id: String,
    pub expected_task_revision: u64,
}

impl TaskMutationInput {
    /// cancel/seen 等 mutation 共用 Thread identity 与 JavaScript-safe revision，不携带本地状态。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        validate_task_revision(&self.task_thread_id, self.expected_task_revision)
    }
}

#[derive(Debug, Clone)]
pub struct TaskTreeDeleteInput {
    pub mutation: TaskMutationInput,
    pub confirm_task_thread_id: String,
}

impl TaskTreeDeleteInput {
    /// 整树删除要求重复精确 identity，防止通用 Thread 删除入口或陈旧选中项误删后代。
    pub(crate) fn validate(&self) -> Result<(), DomainValidationError> {
        self.mutation.validate()?;
        if self.confirm_task_thread_id != self.mutation.task_thread_id {
            return Err(DomainValidationError);
        }
        Ok(())
    }
}

/// 所有 task projection CAS 入口复用同一 identity/整数边界，避免各命令产生细微漂移。
fn validate_task_revision(
    task_thread_id: &str,
    expected_task_revision: u64,
) -> Result<(), DomainValidationError> {
    if !valid_protocol_id(task_thread_id, "thr_", 100)
        || expected_task_revision > MAX_SAFE_JSON_INTEGER
    {
        return Err(DomainValidationError);
    }
    Ok(())
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskSummary {
    pub task_thread_id: String,
    pub parent_thread_id: String,
    pub root_thread_id: String,
    pub origin_turn_id: Option<String>,
    pub task_name: String,
    pub depth: u8,
    pub task_kind: String,
    pub lifecycle: String,
    pub state: String,
    pub revision: u64,
    pub latest_activity_sequence: u64,
    pub unread_count: u64,
    pub descendant_count: u8,
    pub running_descendant_count: u8,
    pub needs_attention_count: u8,
    pub latest_safe_summary: Option<String>,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskActivity {
    pub activity_sequence: u64,
    pub activity_id: String,
    pub task_thread_id: String,
    pub actor_thread_id: String,
    pub causal_turn_id: Option<String>,
    pub kind: String,
    pub summary: String,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskMailboxMessage {
    pub mailbox_sequence: u64,
    pub message_id: String,
    pub sender_thread_id: String,
    pub target_thread_id: String,
    pub causal_turn_id: Option<String>,
    pub kind: String,
    pub content: Vec<TurnContentPart>,
    pub state: String,
    pub bound_turn_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub consumed_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskContextSeed {
    pub context_seed_id: String,
    pub parent_revision: u64,
    pub inheritance_mode: String,
    pub task_brief: Vec<TurnContentPart>,
    pub inherited_context_summary: Option<String>,
    pub inherited_context_preview: Vec<TaskContextPreviewItem>,
    pub fingerprint: String,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskContextPreviewItem {
    pub role: String,
    pub text: Option<String>,
    pub attachment_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskCreateResult {
    pub task: TaskSummary,
    pub turn_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskListResult {
    pub items: Vec<TaskSummary>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskReadResult {
    pub task: TaskSummary,
    pub context_seed: TaskContextSeed,
    pub activities: Vec<TaskActivity>,
    pub mailbox: Vec<TaskMailboxMessage>,
    pub next_cursor: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskObserveResult {
    pub observation_id: String,
    pub task_thread_id: String,
    pub revision: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskMessageResult {
    pub message_id: String,
    pub mailbox_sequence: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskFollowupResult {
    pub message_id: String,
    pub turn_id: String,
    pub task: TaskSummary,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskTreeDeleteResult {
    pub deleted_task_count: u8,
}
